File: python/200.py
# 岛屿数量
# Number of Islands

from common import List


class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        n = len(grid)
        m = len(grid[0])
        islands = 0

        def dfs(i: int, j: int):
            if i < 0 or i >= n or j < 0 or j >= m or grid[i][j] == "0":
                return
            grid[i][j] = "0"
            dfs(i + 1, j)
            dfs(i - 1, j)
            dfs(i, j + 1)
            dfs(i, j - 1)

        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == "1":
                    dfs(i, j)
                    islands += 1

        return islands


if __name__ == "__main__":
    test_cases = [
        [
            ["1", "1", "1", "1", "0"],
            ["1", "1", "0", "1", "0"],
            ["1", "1", "0", "0", "0"],
            ["0", "0", "0", "0", "0"],
        ],
        [
            ["1", "1", "0", "0", "0"],
            ["1", "1", "0", "0", "0"],
            ["0", "0", "1", "0", "0"],
            ["0", "0", "0", "1", "1"],
        ],
    ]
    sol = Solution()
    for case in test_cases:
        res = sol.numIslands(case)
        print(res)

File: cpp/200.cpp
// 岛屿数量
// Number of Islands

#include "common.h"

class Solution
{
public:
    int numIslands(std::vector<std::vector<char>> &grid)
    {
        int n = grid.size();
        int m = grid[0].size();
        int islands = 0;

        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < m; j++)
            {
                if (grid[i][j] == '1')
                {
                    dfs(i, j, n, m, grid);
                    islands++;
                }
            }
        }
        return islands;
    }

    void dfs(int i, int j, int n, int m, std::vector<std::vector<char>> &grid)
    {
        if (i < 0 || i >= n || j < 0 || j >= m || grid[i][j] == '0')
        {
            return;
        }
        grid[i][j] = '0';
        dfs(i + 1, j, n, m, grid);
        dfs(i - 1, j, n, m, grid);
        dfs(i, j + 1, n, m, grid);
        dfs(i, j - 1, n, m, grid);
    }
};

int main()
{
    std::vector<std::vector<std::vector<char>>> test_cases = {
        {{'1', '1', '1', '1', '0'},
         {'1', '1', '0', '1', '0'},
         {'1', '1', '0', '0', '0'},
         {'0', '0', '0', '0', '0'}},
        {{'1', '1', '0', '0', '0'},
         {'1', '1', '0', '0', '0'},
         {'0', '0', '1', '0', '0'},
         {'0', '0', '0', '1', '1'}}};

    Solution sol;
    for (auto &test_case : test_cases)
    {
        int res = sol.numIslands(test_case);
        std::cout << res << std::endl;
    }
}
